### PR TITLE
Louise dev

### PR DESCRIPTION
**Pull request by ShuyunLiuRun**
As a user, I want to have an apparent button so that I can click on it and delete an item in the recorder. 

So I add a delete icon in every card which represents an item in the recorder. When the user clicks on it, there will be a dialogue saying that are you sure you want to delete this item. If the user chooses yes, the item will be deleted from the database and the website. The page will refresh after the dialogue close.

### CODE REVIEW

File: app/src/App.js [hunk 1]
```diff
@@ -1,99 +1,146 @@
...
   const [data, setData] = useState(fd);
-  const [currentContainer, setCurrentContainer] = useState(' ');
+  const [containerLabel, setContainerLabel] = useState(' ');
   const [isLoading, setIsLoading] = useState(true);
   const [init, setInit] = useState(true);
-  const [currentContainerId, setCurrentContainerId] = useState();
+  const [currentContainerId, setCurrentContainerId] = useState(0);
+
+  //initialize the program
+  if (init === true) {
```
Comment (kamifoo) [inline app/src/App.js]:
```suggestion
  if (init) {
```

File: app/src/App.js [hunk 2]
```diff
@@ -1,99 +1,146 @@
...
       alert(error);
       setInit(false)
     });
   }
 
   //once the user click on an item(container), 
   //get all the items stored in this item(container)
-  const clickOnItem = async (name, ID, level, contained_by, additional_json) => {
+  const clickOnItem = async (name, ID, level, contained_by, additional_json, is_container) => {
     console.log(name + ID);
```
Comment (kamifoo) [inline app/src/App.js]:
```suggestion
   
```

File: app/src/App.js [hunk 3]
```diff
@@ -1,99 +1,146 @@
...
-
-    var baseUrl = "http://localhost:4000/";
-    let url = baseUrl + ID;
-
-    Fetch.get(url).then((response) => {
+    setContainerLabel(name);
+    setCurrentContainerId(ID);
+    // if this item is a container, get it's children
+    Fetch.get(baseUrl + ID).then((response) => {
       if (response) {
```
Comment (kamifoo) [inline app/src/App.js:52]: Do we need any specific checking here?

File: app/src/App.js [hunk 4]
```diff
@@ -1,99 +1,146 @@
...
       }
     });
   };
 
+  //Collect new item's data from the form 
+  const handleFormSubmit = (dataFromForm) => {
+    setIsLoading(true);
+    //add container Id for this new item
+    dataFromForm["contained_by"] = currentContainerId;
+    console.log("new item: " + JSON.stringify(dataFromForm));
```
Comment (kamifoo) [inline app/src/App.js]:
```suggestion
    
```

File: app/src/App.js [hunk 5]
```diff
@@ -1,99 +1,146 @@
...
 import * as Fetch from './components/DataComponent.js';
+import AddItemForm from './components/Item/AddItemForm.js';
+import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
+
 
 //fake data
 const fd = require('./data.json');
 
-
+const baseUrl = "http://localhost:4000/";
```
Comment (kamifoo) [inline app/src/App.js]:
```suggestion
const BaseUrl = "http://localhost:4000/";
const ApiCommander = {
   AddItem: ()=>{return BaseUrl + 'addItem/';},
   GetContainerInfo: (containerId) => {return BaseUrl + 'containerInfo/' + containerId},
   //...
}
```

File: app/src/App.js [hunk 6]
```diff
@@ -1,99 +1,146 @@
...
+  if (init === true) {
+    programInit();
+  }
 
-  //the initial data
-  if(init === true){
-    Fetch.get("http://localhost:4000/").then((res) => {
+  // the function to load the initial data
+  function programInit() {
+    Fetch.get(baseUrl).then((res) => {
```
Comment (kamifoo) [inline app/src/App.js]: Just wondering what can we get from base url?

File: app/src/App.js [hunk 7]
```diff
@@ -1,99 +1,146 @@
...
   };
 
+  //Collect new item's data from the form 
+  const handleFormSubmit = (dataFromForm) => {
+    setIsLoading(true);
+    //add container Id for this new item
+    dataFromForm["contained_by"] = currentContainerId;
+    console.log("new item: " + JSON.stringify(dataFromForm));
+    // post the info into db
+    Fetch.post(baseUrl + "addItem/", dataFromForm).then((response) => {
```
Comment (kamifoo) [inline app/src/App.js]:
```suggestion
    Fetch.post(ApiCommander.AddItem(), dataFromForm).then((response) => {
```

File: app/src/App.js [hunk 8]
```diff
@@ -1,99 +1,146 @@
...
-  //   });
-  // };
+  //use the go back button instead of the browsers'
+  const goBack = () => {
+    if (currentContainerId === 1 || containerLabel === ' ') {
+      programInit();
+      setContainerLabel(' ');
+    } else {
+      //get current container's info first
+      Fetch.get(baseUrl + "containerInfo/" + currentContainerId).then((response) => {
```
Comment (kamifoo) [inline app/src/App.js]:
```suggestion
      Fetch.get(ApiCommander.GetContainerInfo(currentContainerId)).then((response) => {
```

File: app/src/components/DataComponent.js [hunk 9]
```diff
@@ -23,10 +22,28 @@ const request = (url, config) => {
             return resJson;
         }
     }).catch((error) => {
-        message.error(error);
+        console.log(error);
     });
 };
 
+// DELETE
+export const de = (url)=>{
```
Comment (kamifoo) [inline app/src/components/DataComponent.js]: Can we please name it as remove()?

File: app/src/App.js (hunk 5 above)
Comment (kamifoo) [reply app/src/App.js]: Oh actually, any reason why we not put this url info in DataComponent.js? As I can see that file is where handling http requests.

File: app/src/components/Item/AddItemForm.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    item_name: "",
+    additional_information: "",
+    is_container: false
+  };
+
+  const handleNameChange = (event) => {
+    data.item_name = event.target.value;
+  }
+
+  const handleAIChange = (event) => {
```
Comment (kamifoo) [inline app/src/components/Item/AddItemForm.js]: We should avoid to use abbreviation to name a method to achieve better readbility.

File: app/src/components/Item/ClickOnOneItem.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import Main from '../Main.js';
+
+
+const ClickOnOneItem = ({clickOnItem, Data, containerLabel, isLoading}) =>{
+    // 这里才去数据库拿东西
+    console.log("containerLabel"+containerLabel);
```
Comment (kamifoo) [inline app/src/components/Item/ClickOnOneItem.js]:
```suggestion

```

File: app/src/components/Item/OneItem.js [hunk 12]
```diff
@@ -19,40 +25,81 @@ const useStyles = makeStyles((theme) => ({
...
+  const { item_name, item_id, level, contained_by, additional_json, is_container } = data;
 
+  //set a state to record status of the dialog
+  const [open, setOpen] = React.useState(false);
 
+  const handleClickOpen = () => {
+    setOpen(true);
+  };
+
+  const handleCloseTrue = () => {
```
Comment (kamifoo) [inline app/src/components/Item/OneItem.js]:
```suggestion
  const handleDialogWhenYes = () => {
```

File: app/src/components/Item/OneItem.js [hunk 13]
```diff
@@ -19,40 +25,81 @@ const useStyles = makeStyles((theme) => ({
...
+    setOpen(false);
+    deleteItem(item_id);
+  };
+
+  const handleCloseFalse = () => {
```
Comment (kamifoo) [inline app/src/components/Item/OneItem.js]: And here

File: server/server.js [hunk 14]
```diff
@@ -1,56 +1,88 @@
...
+
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(bodyParser.json());
 
-app.get('/', async (req, res, next) =>{
-    console.log(req.body);
-    db.query('select * from master where contained_by = 0', [],function(result,fields){
+
+//home page's data
+app.get('/', async (req, res, next) => {
```
Comment (kamifoo) [inline server/server.js]: Please consider to merge this with `/container/:id` get request handler method and leave it returning nothing.

File: server/server.js [hunk 15]
```diff
@@ -1,56 +1,88 @@
...
         console.log(result);
 
-        res.end( JSON.stringify(result));
+        res.end(JSON.stringify(result));
     });
 });
 
-app.get('/:id',function(req,res, next){
+//get specific item's data
+app.get('/:id', function (req, res, next) {
```
Comment (kamifoo) [inline server/server.js]: Can you please change the `/:id` to `/container/:id`? As it is querying container

File: server/server.js [hunk 16]
```diff
@@ -1,56 +1,88 @@
...
-        res.end( JSON.stringify(result));
+        res.end(JSON.stringify(result));
     });
 });
 
-app.post('/addItem', function (req, res) {
-    db.query('select * from master', [],function(result,fields){
-        var toAdd = req.body;
+//get current container info
+app.get('/containerInfo/:id', function (req, res, next) {
```
Comment (kamifoo) [inline server/server.js]: Please consider to change it as `/item/:id` as here querying item id

File: server/server.js [hunk 17]
```diff
@@ -1,56 +1,88 @@
...
- app.post('/container', function (req, res) {
-    db.query('select * from master', [],function(result,fields){
-        var toAdd = req.body;
-        console.log('查询结果：');
+//add new item
+app.post('/addItem', function (req, res) {
+    console.log("request body:")
+    console.table(req.body);
+    var name = req.body.item_name;
+    var adInfo = req.body.additional_information;
```
Comment (kamifoo) [inline server/server.js]: Please change these variable name, `adInfo`, `isC`, `cB`

File: server/server.js [hunk 18]
```diff
@@ -1,56 +1,88 @@
...
+    db.query("INSERT INTO master (item_name, contained_by, is_container, additional_json) VALUES (?,?,?,?);", [name, cB, isC, adInfo], function (result, fields) {
+        console.log('查询Added Item：');
         console.log(result);
 
+        res.end(JSON.stringify(result));
+    });
+});
+
+// delete item
+app.delete('/deleteItem/:id', function (req, res) {
```
Comment (kamifoo) [inline server/server.js]: Also, you can handle delete request by `/item/:id` url as well.
```suggestion
app.delete('/item/:id', function (req, res) {
```

File: server/server.js [hunk 19]
```diff
@@ -1,56 +1,88 @@
...
     });
- });
+});
 
- app.post('/container', function (req, res) {
-    db.query('select * from master', [],function(result,fields){
-        var toAdd = req.body;
-        console.log('查询结果：');
+//add new item
+app.post('/addItem', function (req, res) {
```
Comment (kamifoo) [inline server/server.js]: Please use `/item` url with post request
```suggestion
app.post('/item', function (req, res) {
```

**Review by kamifoo — CHANGES_REQUESTED** [review]

**Comment by kamifoo** [comment]
Empty GoBack.js has been deleted, please refer to commit tab

**Review by kamifoo — APPROVED** [review]
Cheers

**Comment by ShuyunLiuRun** [comment]
Delete Item works.

**Pull request merged**